include/mfmg/hierarchy.hpp

```diff
@@ -0,0 +1,143 @@
+ * This file is part of the mfmg libary. mfmg is distributed under op BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#ifndef MFMG_HIERARCHY_HPP
+#define MFMG_HIERARCHY_HPP
+
+#include <memory>
+#include <vector>
+
+#include <boost/property_tree/ptree.hpp>
+
+namespace mfmg
+{
+
+template <class Operator>
+class Level
+{
+public:
+  std::shared_ptr<const Operator> get_operator() const { return _operator; }
+  std::shared_ptr<const Operator> get_restrictor() const { return _restrictor; }
+  std::shared_ptr<const Operator> get_prolongator() const
+  {
+    return _prolongator;
+  }
+  std::shared_ptr<const Operator> get_smoother() const { return _smoother; }
+
+  void set_operator(std::shared_ptr<const Operator> op) { _operator = op; }
+  void set_restrictor(std::shared_ptr<const Operator> r) { _restrictor = r; }
+  void set_prolongator(std::shared_ptr<const Operator> p) { _prolongator = p; }
+  void set_smoother(std::shared_ptr<const Operator> s) { _smoother = s; }
+
+private:
+  std::shared_ptr<const Operator> _operator, _prolongator, _restrictor,
+      _smoother;
+};
+
+template <class MeshEvaluator, class Operator, class Vector>
+class Hierarchy
+{
+public:
+  using mesh_type = typename MeshEvaluator::mesh_type;
+  using vector_type = Vector;
+  using operator_type = Operator;
+
+public:
+  Hierarchy(MPI_Comm comm, MeshEvaluator &evaluator, mesh_type &mesh,
+            std::shared_ptr<boost::property_tree::ptree> params = nullptr)
+  {
+    static_assert(
+        std::is_same<typename MeshEvaluator::operator_type, Operator>::value,
+        "Operator template does not match the one in MeshEvaluator");
+    static_assert(
+        std::is_same<typename MeshEvaluator::vector_type, Vector>::value,
+        "Vector template does not match the one in MeshEvaluator");
+
+    const int num_levels = 2;
+    _levels.resize(2);
+
+    _levels[0].set_operator(evaluator.get_operator(mesh));
+    for (int i = 0; i < num_levels; i++)
+    {
+      auto level_prev = _levels[i];
+      auto level_coarse = _levels[i + 1];
+
+      auto a = level_prev.get_operator();
+
+      auto restrictor = Adapter<MeshEvaluator, Operator>::build_restrictor(
+          comm, evaluator, mesh, params);
+
+      auto prolongator = restrictor->transpose();
+      auto ap = a->multiply(*prolongator);
+      auto a_coarse = restrictor->multiply(*ap);
+
+      level_coarse.set_restrictor(restrictor);
+      level_coarse.set_prolongator(prolongator);
+      level_coarse.set_operator(a_coarse);
+
+      // auto smoother = details::build_smoother(op);
+    }
+  }
+
+  void apply(const Vector &b, Vector &x, int start_level = 0) const
```

Comment:
So how this supposed to be called? Our first goal is to use deal, so we need to have a function `vmult( Vector &x, const Vector &b)` How do pass the level?

Reply:
I'm adding something like
```c++ 
  void vmult(Vector &x, const Vector &b) const { apply(b, x, 0); }
```
to be used with deali.ii.